### PR TITLE
Convert timestamp methods on `DateTime` to return `Result`s

### PR DESCRIPTION
**Pull request by pitdicker**
All the preparations are in!
Also adds an `ok_or!` macro.

cc @Zomtir.

### CODE REVIEW

File: src/datetime/tests.rs [hunk 1]
```diff
@@ -217,15 +219,15 @@ fn test_datetime_from_timestamp_nanos() {
     let nanos = parsed.timestamp_nanos().unwrap();
     assert_eq!(
         Some(DateTime::from_timestamp_nanos(nanos)),
```
Comment (Zomtir) [inline src/datetime/tests.rs:221]: I'd prefer to do the comparison in the `Result` domain:

```
    assert_eq!(
        DateTime::from_timestamp_nanos(nanos),
        DateTime::from_timestamp(nanos / A_BILLION, (nanos % A_BILLION) as u32)
    );
```

File: src/datetime/mod.rs [hunk 2]
```diff
@@ -282,7 +282,8 @@ impl<Tz: TimeZone> DateTime<Tz> {
             subsec_nanos -= 1_000_000_000;
             timestamp += 1;
         }
-        try_opt!(timestamp.checked_mul(1_000_000_000)).checked_add(subsec_nanos)
+        let ts = try_err!(ok_or!(timestamp.checked_mul(1_000_000_000), Error::OutOfRange));
```
Comment (Zomtir) [inline src/datetime/mod.rs:285]: If I remember correctly, we want the value to be unwrapped anyway in all instances. I sugges `try_ok_or` instead that directly unwraps T or returns the error early.

`try_err!(ok_or!(..))` vs `try_ok_or!`

I don't really mind chaining macros, but wanted to have it mentioned.

**Review by Zomtir — APPROVED** [review]

**Review by djc — APPROVED** [review]

**Comment by pitdicker** [comment]
Updated to have a `to_or` macro and used `?` in the doctests (starting to get payoffs :smile:).

**Comment by pitdicker** [comment]
Sorry, just after your review.

It is a combination of `try_err!(ok_or!(..))`.

**Comment by djc** [comment]
I don't like the `to_or!()` name -- `try_ok_or!()` looks better to me.

File: src/datetime/tests.rs [hunk 3]
```diff
@@ -234,7 +236,7 @@ fn test_datetime_from_timestamp_nanos() {
     for secs in secs_test.iter().copied() {
         assert_eq!(
             Some(DateTime::from_timestamp_nanos(secs * 1_000_000_000)),
-            DateTime::from_timestamp(secs, 0)
+            DateTime::from_timestamp(secs, 0).ok()
```
Comment (Zomtir) [inline src/datetime/tests.rs:239]: This one slipped through.

Comment (pitdicker) [reply src/datetime/tests.rs]: Oops, thank you.

**Comment by pitdicker** [comment]
> I don't like the `to_or!()` name -- `try_ok_or!()` looks better to me.

Yea, that is better.

**Review by djc — APPROVED** [review]

**Pull request merged**